From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Tue, 14 Oct 2025 15:19:21 +0800
Subject: [PATCH] ntdll: Implement NtCancelWaitCompletionPacket().

(cherry picked from commit 8809cc80f11196616bf2ff7d19aa50eee09dfc68)

diff --git a/dlls/ntdll/ntdll.spec b/dlls/ntdll/ntdll.spec
index 489637433ab..7689f5abf70 100644
--- a/dlls/ntdll/ntdll.spec
+++ b/dlls/ntdll/ntdll.spec
@@ -170,6 +170,7 @@
 @ stdcall -syscall NtCancelIoFileEx(long ptr ptr)
 @ stdcall -syscall NtCancelSynchronousIoFile(long ptr ptr)
 @ stdcall -syscall=0x0061 NtCancelTimer(long ptr)
+@ stdcall -syscall NtCancelWaitCompletionPacket(ptr long)
 @ stdcall -syscall=0x003e NtClearEvent(long)
 @ stdcall -syscall=0x000f NtClose(long)
 @ stdcall -syscall=0x003b NtCloseObjectAuditAlarm(ptr long long)
diff --git a/dlls/ntdll/signal_arm64ec.c b/dlls/ntdll/signal_arm64ec.c
index fe219ab4bf6..1fa3fcf7262 100644
--- a/dlls/ntdll/signal_arm64ec.c
+++ b/dlls/ntdll/signal_arm64ec.c
@@ -379,6 +379,7 @@ DEFINE_SYSCALL(NtCancelIoFile, (HANDLE handle, IO_STATUS_BLOCK *io_status))
 DEFINE_SYSCALL(NtCancelIoFileEx, (HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status))
 DEFINE_SYSCALL(NtCancelSynchronousIoFile, (HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status))
 DEFINE_SYSCALL(NtCancelTimer, (HANDLE handle, BOOLEAN *state))
+DEFINE_SYSCALL(NtCancelWaitCompletionPacket, (HANDLE packet, BOOLEAN remove_signaled))
 DEFINE_SYSCALL(NtClearEvent, (HANDLE handle))
 DEFINE_SYSCALL(NtClose, (HANDLE handle))
 DEFINE_SYSCALL(NtCloseObjectAuditAlarm, (UNICODE_STRING *subsystem, HANDLE handle, BOOLEAN onclose))
diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index e7d02576ed0..9fab4e18ec1 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -8332,7 +8332,6 @@ static void test_cancel_wait_completion_packet(void)
 
     if (!pNtCancelWaitCompletionPacket)
     {
-        todo_wine
         win_skip("NtCancelWaitCompletionPacket is unavailable.\n");
         return;
     }
diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index 0bc04f94388..07a67773c0e 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -3716,6 +3716,26 @@ NTSTATUS WINAPI NtAssociateWaitCompletionPacket( HANDLE packet, HANDLE completio
     return ret;
 }
 
+/***********************************************************************
+ *           NtCancelWaitCompletionPacket (NTDLL.@)
+ */
+NTSTATUS WINAPI NtCancelWaitCompletionPacket( HANDLE packet, BOOLEAN remove_signaled )
+{
+    NTSTATUS status;
+
+    TRACE( "%p, %d\n", packet, remove_signaled );
+
+    SERVER_START_REQ( cancel_completion_packet )
+    {
+        req->packet = wine_server_obj_handle( packet );
+        req->remove_signaled = remove_signaled;
+        status = wine_server_call( req );
+    }
+    SERVER_END_REQ;
+    return status;
+}
+
+
 /***********************************************************************
  *           NtCreateWaitCompletionPacket (NTDLL.@)
  */
diff --git a/dlls/wow64/sync.c b/dlls/wow64/sync.c
index 2fef74d90ab..861e578fcb1 100644
--- a/dlls/wow64/sync.c
+++ b/dlls/wow64/sync.c
@@ -1907,6 +1907,18 @@ NTSTATUS WINAPI wow64_NtAssociateWaitCompletionPacket( UINT *args )
 }
 
 
+/**********************************************************************
+ *           wow64_NtCancelWaitCompletionPacket
+ */
+NTSTATUS WINAPI wow64_NtCancelWaitCompletionPacket( UINT *args )
+{
+    HANDLE packet = get_handle( &args );
+    BOOLEAN remove_signaled = get_ulong( &args );
+
+    return NtCancelWaitCompletionPacket( packet, remove_signaled );
+}
+
+
 /**********************************************************************
  *           wow64_NtCreateWaitCompletionPacket
  */
diff --git a/include/winternl.h b/include/winternl.h
index 678bd574611..442b574a07b 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -4811,6 +4811,7 @@ NTSYSAPI NTSTATUS  WINAPI NtCancelIoFile(HANDLE,PIO_STATUS_BLOCK);
 NTSYSAPI NTSTATUS  WINAPI NtCancelIoFileEx(HANDLE,PIO_STATUS_BLOCK,PIO_STATUS_BLOCK);
 NTSYSAPI NTSTATUS  WINAPI NtCancelSynchronousIoFile(HANDLE,PIO_STATUS_BLOCK,PIO_STATUS_BLOCK);
 NTSYSAPI NTSTATUS  WINAPI NtCancelTimer(HANDLE, BOOLEAN*);
+NTSYSAPI NTSTATUS  WINAPI NtCancelWaitCompletionPacket(HANDLE, BOOLEAN);
 NTSYSAPI NTSTATUS  WINAPI NtClearEvent(HANDLE);
 NTSYSAPI NTSTATUS  WINAPI NtClose(HANDLE);
 NTSYSAPI NTSTATUS  WINAPI NtCloseObjectAuditAlarm(PUNICODE_STRING,HANDLE,BOOLEAN);
diff --git a/server/completion.c b/server/completion.c
index a0aa38b793d..ebce2ab66a7 100644
--- a/server/completion.c
+++ b/server/completion.c
@@ -695,3 +695,30 @@ DECL_HANDLER(associate_completion_packet)
     release_object( target );
     release_object( packet );
 }
+
+/* cancel a wait completion packet */
+DECL_HANDLER(cancel_completion_packet)
+{
+    struct completion_packet *packet;
+
+    packet = get_completion_packet_obj( current->process, req->packet, WAIT_COMPLETION_PACKET_QUERY_STATE );
+    if (!packet)
+        return;
+
+    if (!packet->in_target_packet_queue && !packet->in_completion_queue)
+    {
+        set_error( STATUS_CANCELLED );
+        release_object( packet );
+        return;
+    }
+
+    if (packet->in_completion_queue && !req->remove_signaled)
+    {
+        set_error( STATUS_PENDING );
+        release_object( packet );
+        return;
+    }
+
+    cancel_completion_packet( packet );
+    release_object( packet );
+}
diff --git a/server/protocol.def b/server/protocol.def
index 751275a70cb..0d19ee49ef8 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -4018,6 +4018,13 @@ struct handle_info
 @END
 
 
+/* Cancel a wait completion packet */
+@REQ(cancel_completion_packet)
+    obj_handle_t  packet;         /* wait completion packet handle */
+    int           remove_signaled;/* whether to remove signaled packet */
+@END
+
+
 /* check for associated completion and push msg */
 @REQ(add_fd_completion)
     obj_handle_t   handle;        /* async' object */
